### DISCUSSION TITLE
Declaration for Generic Function Returning a ZodObject

### DISCUSSION THREAD

**Discussion by BodneyC** [colinhacks/zod#1827] [Q&A]
Hello,

## The Issue

I'm currently facing errors in the declaration generated for a generic function returning a `ZodObject` in which the generic, a `T extends ZodTypeAny`, is used. It follows the pattern laid out [in the readme](https://github.com/colinhacks/zod#writing-generic-functions).

The actual function works fine, it's just the declaration causing me problems when using this project in another (see below for details)

## Minimal Reproducible Example

```ts
// test.ts
import { z } from 'zod';
export const a = <T extends z.ZodTypeAny>(b: T) => z.object({ someKey: b });
```

```ts
// package.json
{
  "scripts": { "build": "rm -rf dist && tsc -b tsconfig.json" },
  "dependencies": { "zod": "^3.20.2" },
  "devDependencies": { "typescript": "^4.9.4" }
}
```

```ts
// tsconfig.json
{
  "compilerOptions": {
    "target": "es2016",
    "module": "commonjs",
    "declaration": true,
    "outDir": "dist",
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "skipLibCheck": true
  }
}
```

Then `npm run build` generates `./dist/test.d.ts`

## The Error

```ts
// ./dist/test.d.ts
import { z } from 'zod';
export declare function a<T extends z.ZodTypeAny>(b: T): z.ZodObject<{
    someKey: T;
}, "strip", z.ZodTypeAny, z.objectUtil.addQuestionMarks<{
    someKey: T["_output"];
}> extends infer T_1 ? { [k_1 in keyof T_1]: z.objectUtil.addQuestionMarks<{
    someKey: T["_output"];
}>[k_1]; } : never, z.objectUtil.addQuestionMarks<{
    someKey: T["_input"];
}> extends infer T_2 ? { [k_3 in keyof T_2]: z.objectUtil.addQuestionMarks<{
    someKey: T["_input"];
}>[k_3]; } : never>;
```

Both `extends infer...` lines give the error:

`Type 'k_x' cannot be used to index type 'addQuestionMarks<{ someKey: T["_output"]; }>'. [2536]`

## Finally

Its thoroughly possible I've missed something obvious so please let me know if that's the case

If not, then any workaround you can think of would be appreciated

**Comment by JacobWeisenburger** [comment]
Does this work for you?

```ts
declare function a<T extends z.ZodTypeAny> ( b: T ): z.ZodObject<{
    someKey: T
}, "strip", z.ZodTypeAny, {
    someKey: T[ '_output' ]
}, {
    someKey: T[ '_output' ]
}>
```

**Comment by BodneyC** [comment]
Yes, if I delete the `extends infer...` chunks of the declaration then it works without a hitch. Though I'm unsure of how to get `tsc` to generate the declaration like that on build, or would this be a post-build edit as a workaround?

I could add that return type to the `.ts` however the real object in question is significantly larger and may become a little unwieldy/tough to maintain

**Comment by JacobWeisenburger** [comment]
This seems like it's not really a zod issue. Is it ok if I move this to the discussions?

**Comment by BodneyC** [comment]
If that's a better fit then sure - I was a little undecided on where to raise this issue to begin with

**Comment by sinuheshinbr** [comment]
Any updates on this? I'm having the same problem, what's the best approach here?
